### PR TITLE
Nonlinear preconditioning

### PR DESCRIPTION
**Pull request by danshapero**
Firedrake can use PETSc's nonlinear preconditioning features through the `AuxiliaryOperatorSNES` class. This patch adds a class `IRKAuxiliarySNES` which is almost identical to `IRKAuxiliaryOperatorPC` so that Irksome can use nonlinear preconditioning.

I added a test case based on the Stefan problem. This is similar to the test case for lagging that I used in #228. But now the lagged form is a preconditioner to the fully implicit form.

### CODE REVIEW

File: tests/test_npc.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    }
+    method = BackwardEuler()
+    stepper = TimeStepper(F, method, t, dt, T, **params)
+
+    final_time = 20.0
+    num_steps = int(final_time / float(dt))
+    for _ in range(num_steps):
+        stepper.advance()
+        t.assign(t + dt)
+        assert stepper.solver.snes.getConvergedReason() > 0
```
Comment (pbrubeck) [inline tests/test_npc.py]: We should assert that we get a reasonably accurate solution or conservation/dissipation of some quantity

Comment (danshapero) [reply tests/test_npc.py]: I changed this to assert that it converged due to FNORM_RELATIVE. The boundary conditions are oscillating, so unlike the case of running to steady state there isn't an obvious dissipation potential. If I change it to run to steady state then you get into annoying issues of setting an absolute tolerance. If you think it's better to use this as a demo than a unit test then I can think of something else to try instead. I just wanted something that will obviously break Newton.

Comment (rckirby) [reply tests/test_npc.py]: I think breaking/not breaking Newton as expected as a test is fine.  But, since this is a new feature, if you're offering to write a demo, we'd gladly take it.  It could be a separate PR, and might take some text explaining what the nonlinear PC is doing, etc.

**Review by rckirby — APPROVED** [review]
I'm happy to approve this.  I'd like to eventually have a demo, but that could be a separate, later contribution.

File: irksome/pc.py [hunk 2]
```diff
@@ -81,6 +81,65 @@ def form(self, pc, test, trial):
         return Jnew, bcnew
 
 
+class IRKAuxiliaryOperatorSNES(AuxiliaryOperatorSNES):
+    """Base class that inherits from Firedrake's :class:`AuxiliaryOperatorSNES`
```
Comment (JHopeCollins) [inline irksome/pc.py]: @pbrubeck will this resolve to link to the Firedrake docs? Does Irksome have the intersphinx set up?
Even with intersphinx it might have to be `:class:`~firedrake.AuxiliaryOperatorSNES`` to resolve correctly.

File: irksome/pc.py [hunk 3]
```diff
@@ -81,6 +81,65 @@ def form(self, pc, test, trial):
...
+        """Derived classes produce a typically structured
+        approximation to A."""
+        raise NotImplementedError
+
+    def form(self, snes, w0, w, test):
+        """Implements the interface for AuxiliaryOperatorSNES.
+
+        :arg snes: the PETSc SNES object.
+        :arg w0: the current iterate of the stages (unused by default, but
+            available to subclasses that wish to lag terms).
```
Comment (JHopeCollins) [inline irksome/pc.py:161]: @pbrubeck whats the state of https://github.com/firedrakeproject/Irksome/pull/156? We'd need that functionality to properly use `w0` here.

If that PR isn't ready soon then we can at least raise an issue to fix this SNES once that PR is merged.

File: tests/test_npc.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            },
+        },
+    }
+    method = BackwardEuler()
+    stepper = TimeStepper(F, method, t, dt, T, **params)
+
+    reasons = [
+        PETSc.SNES.ConvergedReason.CONVERGED_FNORM_ABS,
+        PETSc.SNES.ConvergedReason.CONVERGED_FNORM_RELATIVE,
+        PETSc.SNES.ConvergedReason.CONVERGED_SNORM_RELATIVE,
```
Comment (JHopeCollins) [inline tests/test_npc.py]: Are we ok with stagnation here? We also aren't setting the `snes_atol` in the options dictionary so testing that convergence isn't robust to future changes.
If you set `snes_stol: 0` and an explicit `snes_atol` (possibly also 0) then you can just check the solver didn't diverge.

File: irksome/pc.py (hunk 2 above)
Comment (danshapero) [reply irksome/pc.py]: The only way that I can get this to work is with the fully-qualified name:
```
:class:`~firedrake.preconditioners.auxiliary_snes.AuxiliaryOperatorSNES`
```
But Sphinx automatically generates a correct link to AuxiliaryOperatorSNES right above the docstring itself because it's a base class. It seems redundant to have two links to the same class all within a paragraph of each other in the docstring, so I changes those links to regular text.

File: tests/test_npc.py (hunk 4 above)
Comment (danshapero) [reply tests/test_npc.py]: I tried `snes_stol: 0` and this test looks a bit worse now. Most of the time it converges due to fnorm_relative. Every so often the solver will stagnate but then go back to normal at the next timestep. I wasn't able to get rid of the stagnation by messing with solver options. I get the same behavior even if I do the timestepping manually and use the previous guess of the solver (i.e. true Picard iteration) as a NPC. I can work up another test case using Allen-Cahn or something we already know if we're not ok with occasional stagnation.

Comment (JHopeCollins) [reply tests/test_npc.py]: Personally I would prefer not equating stagnating with converging, so if there is an alternative I'd prefer that. But I am not familiar at all with this equation so maybe there is a detail here I'm missing. If so then it'd be useful to have a quick comment next to the `reasons` to explain why we're ok with snorm sometimes.

File: irksome/pc.py (hunk 3 above)
Comment (pbrubeck) [reply irksome/pc.py:161]: The PR is stale. I remember things were in a good state when I last worked on it.

File: tests/test_npc.py (hunk 4 above)
Comment (danshapero) [reply tests/test_npc.py]: I changed the test case to use a very sharp but smooth transition between the solid and liquid phases. It breaks Newton but works with the lagged NPC.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -15,9 +15,12 @@ concurrency:
 
 jobs:
   test:
+    strategy:
+      matrix:
+        version: [2026.4.0, dev-release]
```
Comment (JHopeCollins) [inline .github/workflows/ci.yml]: Saves having to update this number each time there's a new release.
```suggestion
        version: [latest, dev-release]
```
